Improve documentation for bitops

konsumlamm opened the pull request:
* improve wording
* use `func` (instead of `proc` with `{.noSideEffect.}`)
* use `let` in `runnableExamples` where possible
* add some links

EDIT: Closes #7587.

konsumlamm committed:
Improve documentation for bitops

Use func
Use let in runnableExamples

timotheecour commented:
let's wait for https://github.com/nim-lang/Nim/pull/16622 to get merged 1st as that PR has been in the queue for a while, to avoid conflicts there

konsumlamm committed:
Remove unnecessary tests

Fix #7587

konsumlamm commented:
* all the tests in `main2` were already in `main1`, so i renamed `main1` to `main` and removed `main2`
* add a test as suggested in https://github.com/nim-lang/Nim/issues/7587#issuecomment-629617266

timotheecour commented:
@konsumlamm ok https://github.com/nim-lang/Nim/pull/16622 got merged, please rebase :)

konsumlamm committed:
Merge branch 'devel' of https://github.com/nim-lang/Nim into bitops

timotheecour commented on lib/pure/bitops.nim:
```diff
@@ -479,24 +483,24 @@ when useGCC_builtins:
-      importc: "__popcnt", header: "<intrin.h>", noSideEffect.}
-  proc builtin_popcnt64(a2: uint64): uint64 {.
-      importc: "__popcnt64", header: "<intrin.h>", noSideEffect.}
+  func builtin_popcnt16(a2: uint16): uint16 {.
```
optional: turn into more readable 1-liners?
```nim
func builtin_popcnt16(a2: uint16): uint16 {.importc: "__popcnt16", header: "<intrin.h>".}
```
(even if exceeds that (IMO silly) 80 col limit a bit)

timotheecour replied:
let's defer this to an eventual followup PR instead

timotheecour commented on lib/pure/bitops.nim:
```diff
@@ -562,13 +566,13 @@ proc countSetBits*(x: SomeInteger): int {.inline, noSideEffect.} =
-proc parityBits*(x: SomeInteger): int {.inline, noSideEffect.} =
-  ## Calculate the bit parity in integer. If number of 1-bit
-  ## is odd parity is 1, otherwise 0.
+func parityBits*(x: SomeInteger): int {.inline.} =
```
pre-existing, but why isn't this returning `bool`? nim isn't C...; casting should have no cost, and resulting api would be easier; maybe we need another proc that returns bool (maybe `bitParity` ?)

timotheecour commented on lib/pure/bitops.nim:
```diff
@@ -589,10 +593,10 @@ proc parityBits*(x: SomeInteger): int {.inline, noSideEffect.} =
-  ## otherwise result is undefined.
+func firstSetBit*(x: SomeInteger): int {.inline.} =
+  ## Returns the 1-based index of the least significant set bit of `x`.
+  ## If `x` is zero, when `noUndefinedBitOpts` is set, the result is 0,
```
pre-existing but IMO future PR should add `-d:nimNoUndefinedBitOpts` of which noUndefinedBitOpts would become an undocumented alias

timotheecour reviewed (approved):
LGTM, I've checked everything including fact that large `main2` block was entirely covered by other tests block even though that's not obvious at 1st sight: some static blocks in main2 are covered by fact we have `static: main`

the duplication was introduced in https://github.com/nim-lang/Nim/pull/13803 (note that before https://github.com/nim-lang/Nim/pull/13803 bitops were also tested with 
-d:noIntrinsicsBitOpts
-d:noUndefinedBitOps
(see https://github.com/nim-lang/Nim/pull/13803/files#diff-cc11adc098ea503d26c0788e763173c275570c6074503ebea374ed348344562c)

in future PR we should add:
```nim
discard """
 matrix: "; -d:noIntrinsicsBitOpts -d:noUndefinedBitOps"
"""
```
to test this as well
